Refactor annotation module and handle if enable_bounding_box turns on

nanli-emory opened the pull request:
1. fixing the numerical precision in 'BaseImage'
2. refactoring the existing 'AnnotationModule'
3. deal with the bounding box in 'AnnotationModule'

nanli-emory committed:
1. fixing the numerical precision in 'BaseImage'
2. refactoring the existing 'AnnotationModule'
3. deal with bounding box in 'AnnotationModule'

nanli-emory committed:
add "_annot_json.png" and "_annot_xml.png" into  'global_config.js' files to show up the annotation mask images.

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -197,7 +197,7 @@ def getImgThumb(self, dim):
                         if self["enable_bounding_box"]
                         else osh.level_dimensions[level])
-                if relative_down == 1.0: #there exists an open slide level exactly for this requested mag
+                if np.isclose(relative_down, 1, atol=.01): #there exists an open slide level exactly for this requested mag
```
A thought for you, perhaps "getBestLevelForDownsample" should return a binary flag as well to indicate if the level proposed is an existing openslide level?

in which case, this if statement would check the value of that boolean variable, instead of essentially doing the same computation again?

nanli-emory replied:
Hi @choosehappy, I added a flag to indicate if the level proposed is an existing openslide level.
Also, merged the PR 236 code into BaseImage to fix the incorrect size output.

choosehappy commented on histoqc/AnnotationModule.py:
```diff
@@ -95,74 +96,69 @@ def mask_out_annotation(s, point_sets):
+    if not format:
+        msg = f"format is not provided, using xml as the default format."
+        logging.warning(f"{s['filename']} - {msg}")
+        s["warnings"].append(msg) 
```
its not immediately clear to me how this code works? if format is None, where is it set so that the default becomes XML?

nanli-emory replied:
Yes, you are right. set 'xml' as default.

choosehappy commented on histoqc/AnnotationModule.py:
```diff
@@ -95,74 +96,69 @@ def mask_out_annotation(s, point_sets):
+        msg = f"file path is not provided, using \"{s['dir']}\" as the default file path"
+        logging.warning(f"{s['filename']} - {msg}")
+        s["warnings"].append(msg)        
+        file_path = s["dir"]
```
note here that file_path is set, but in the above if statement format is not set?

nanli-emory committed:
1. set xml as default if format is not provided.
2. add an example of config file for annotation module.
3. remove the previous config files for xml and geojson.

nanli-emory committed:
refactor BaseImage
1. Fixing outputting incorrect image size issue based on PR #236
2. enhanced BaseImage arguments verification
3. enhanced getBestLevelForDownsample returns (level, isExactLevel)
4. add a helper function - getDimensionsByOneDim to get the dims
5. add resizeTileDownward function to handle scaling down the thumb image

nanli-emory committed:
clear up the message

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -139,89 +92,200 @@ def addToPrintList(self, name, val):
-                size=(tuple((np.array(img_base_size)/osh.level_downsamples[level]).astype(int))
+            return (osh.get_best_level_for_downsample(downsample_factor), False)
+
+    # this approach sequentially loads smaller tiles, resizes them, and then merges them together,
```
i think this comment was accidently still leftin place?

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -139,89 +92,200 @@ def addToPrintList(self, name, val):
-                logging.error(
-                    f"{self['filename']}: Unknown image level setting: {dim}!")
-                return -1
+                # recommand having the dimension is less than 10k     
```
recommend : )

choosehappy commented on histoqc/BaseImage.py:
```diff
@@ -139,89 +92,200 @@ def addToPrintList(self, name, val):
+                cloest_region = rgba2rgb(self, cloest_region)
+            target_region = cloest_region.resize((win_down_width, win_down_height))
+            row_piece.append(target_region)
+        row_piece = np.concatenate(row_piece, axis=0)
```
can you adda comment here to note that this isn't very efficient, and if more efficiency isneeded we should likely refactor using "paste" from Image. or even just set the pixels directly with indexing?

nanli-emory committed:
1. remove the useless comments
2. add Notes for future enhance on resizing the tile images.

nanli-emory committed:
fix the typo
